### PR TITLE
Canonicalize set/frozenset iteration order in _stabilize_tensor_subclass_metadata

### LINKED ISSUE #193418: set/frozenset in tensor subclass metadata causes non-deterministic AOTAutograd cache keys
## Bug

`_stabilize_tensor_subclass_metadata` in `torch/_functorch/_aot_autograd/autograd_cache.py` doesn't handle `set` or `frozenset`. They fall through to `return obj` unchanged. When pickled for cache key generation, `pickle.dumps()` writes set elements in hash-table iteration order, which changes across processes due to Python's `PYTHONHASHSEED` randomization.

Same metadata, different processes, different cache keys, **silent cache misses** in the persistent AOTAutograd cache. No wrong results, the cache just never hits across process boundaries.

This is a pre-existing issue on main. It was flagged by @aorenste during review of PR #190080 for Issue #189293 related to `WeakValueDictionary` pickle crash.

## Reproducer

Replicates the exact code path from `AOTAutogradCachePickler`: `__tensor_flatten__()` to `_stabilize_tensor_subclass_metadata()` to `pickle.dumps()` to `blake2b` to cache key.

```python
import hashlib, os, pickle, sys, subprocess

SUBPROCESS = "_REPRO_SUBPROCESS" in os.environ

# Exact logic from AOTAutogradCachePickler (autograd_cache.py lines 725-738 on main)

def _stabilize_tensor_subclass_metadata(obj):
    if isinstance(obj, tuple):
        return tuple(_stabilize_tensor_subclass_metadata(x) for x in obj)
    if isinstance(obj, list):
        return [_stabilize_tensor_subclass_metadata(x) for x in obj]
    if isinstance(obj, dict):
        return {k: _stabilize_tensor_subclass_metadata(v) for k, v in obj.items()}
    return obj  # <-- set/frozenset fall through unchanged

def _default_stable_hash_for_caching(subclass_metadata, shape, requires_grad):
    cache_payload = (
        shape, requires_grad,
        _stabilize_tensor_subclass_metadata(subclass_metadata),
        {},
    )
    return hashlib.blake2b(pickle.dumps(cache_payload), digest_size=16).hexdigest()

if SUBPROCESS:
    metadata = {"string_set": {"alpha", "beta", "gamma", "delta", "epsilon"}}
    h = _default_stable_hash_for_caching(metadata, (3,), False)
    print(f"PYTHONHASHSEED={os.environ.get('PYTHONHASHSEED','random')}  cache_hash={h}")
else:
    script = os.path.abspath(__file__)
    hashes = {}
    for seed in ["42", "123", "999"]:
        env = {**os.environ, "_REPRO_SUBPROCESS": "1", "PYTHONHASHSEED": seed}
        r = subprocess.run([sys.executable, script], env=env, capture_output=True, text=True)
        line = r.stdout.strip()
        print(line)
        hashes[seed] = line.split("cache_hash=")[1]
    unique = set(hashes.values())
    if len(unique) == 1:
        print("\nPASS: cache hash is stable across all 3 processes")
    else:
        print(f"\nFAIL: {len(unique)} different cache hashes for the same tensor")
```

**Output (Python 3.12):**

```text
PYTHONHASHSEED=42   cache_hash=7f90c9a2be29addb00334f2c3aba26f6
PYTHONHASHSEED=123  cache_hash=aa6b8bc80999b969c183dfff76d11c0e
PYTHONHASHSEED=999  cache_hash=ff1b50a1677083877b576d508a6389dc

FAIL: 3 different cache hashes for the same tensor
```

Same bug also affects `frozenset` in metadata.

## Fix

Canonicalize by sorting the pickled bytes of each element:

```python
if isinstance(obj, (set, frozenset)):
    return sorted(
        pickle.dumps(self._stabilize_tensor_subclass_metadata(x))
        for x in obj
    )
```

cc @ezyang @albanD @chauhang @penguinwu @oulgen @jamesjwu @aorenste @anijain2305 @laithsakka @masnesral @coconutruben @aditvenk @bdhirsh @bobrenjc93 @amjames

### Versions

Silent cache misses in AOTAutograd persistent cache when tensor subclass metadata contains set/frozenset, due to non-deterministic pickle iteration order across processes.

### PR DESCRIPTION
**Pull request by Arindamsamanta2004**
Fixes #193418

## Summary

`_stabilize_tensor_subclass_metadata` did not handle `set` or `frozenset`. They fell through to `return obj` unchanged. When pickled for cache key generation, `pickle.dumps()` serializes set elements in hash-table iteration order, which is randomized per process by `PYTHONHASHSEED`. This causes the same tensor subclass metadata to produce different cache keys in different processes, resulting in silent cache misses in the persistent AOTAutograd cache.

The fix canonicalizes set/frozenset by individually pickling each recursively stabilized element and sorting the resulting byte strings. This produces a deterministic `list[bytes]` regardless of hash seed or process.

Flagged by @aorenste during review of PR #190080.

## Test plan

Standalone reproducer confirms identical cache hashes across 3 different PYTHONHASHSEED values:

```
PYTHONHASHSEED=42   cache_hash=d474a5deeeb7bfbe0e04a629c5f4dc68
PYTHONHASHSEED=123  cache_hash=d474a5deeeb7bfbe0e04a629c5f4dc68
PYTHONHASHSEED=999  cache_hash=d474a5deeeb7bfbe0e04a629c5f4dc68
PASS: cache hash is stable across all 3 processes (fix works!)
```

Three new unit tests in `AOTAutogradCachePicklerTests`:
- `test_stabilize_set_deterministic_order`
- `test_stabilize_set_in_dict_deterministic`
- `test_stabilize_nested_frozenset_in_set`

All three pass with the fix and fail without it.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 1]
```diff
@@ -735,6 +735,11 @@ def _stabilize_tensor_subclass_metadata(self, obj: Any) -> Any:
             return {
                 k: self._stabilize_tensor_subclass_metadata(v) for k, v in obj.items()
             }
+        if isinstance(obj, (set, frozenset)):
+            return sorted(
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: [blocker] My agent says:

> Preserve the container kind in a private, typed canonical representation, and add non-collision coverage for `set`, `frozenset`, and ordinary `list[bytes]` metadata.
>
> ```python
> return _CanonicalSetMetadata(
>     container_type=type(obj),
>     elements=tuple(sorted(...)),
> )
> ```
>
> Returning a bare list makes `{x}`, `frozenset({x})`, and `[pickle.dumps(x)]` produce the same cache key. Cache hits reuse serialized `SubclassCreationMeta.meta` without revalidating it, so the wrong metadata can reach `__tensor_unflatten__`.

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

File: torch/_functorch/_aot_autograd/autograd_cache.py (hunk 1 above)
Comment (Arindamsamanta2004) [reply torch/_functorch/_aot_autograd/autograd_cache.py]: Thanks for flagging, I will update the fix to use a _CanonicalSetMetadata namedtuple that preserves the container_type, so set and frozenset produce distinct cache keys.

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 2]
```diff
@@ -735,6 +740,16 @@ def _stabilize_tensor_subclass_metadata(self, obj: Any) -> Any:
             return {
                 k: self._stabilize_tensor_subclass_metadata(v) for k, v in obj.items()
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py:741]: > Future AI reviewer: #190080 already stabilizes dictionary keys. Do not flag that as missing here again.

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 3]
```diff
@@ -735,6 +740,16 @@ def _stabilize_tensor_subclass_metadata(self, obj: Any) -> Any:
...
             }
+        if isinstance(obj, (set, frozenset)):
+            return _CanonicalSetMetadata(
+                container_type=type(obj),
+                elements=tuple(
+                    sorted(
+                        pickle.dumps(self._stabilize_tensor_subclass_metadata(x))
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py:748]: [cleanup] My agent says:

> Preserve the stabilized objects until the outer cache payload is pickled:
>
> ```python
> elements=tuple(
>     sorted(
>         (self._stabilize_tensor_subclass_metadata(x) for x in obj),
>         key=pickle.dumps,
>     )
> ),
> ```
>
> Pickling each element independently resets pickle's memo and loses references shared across elements. A set whose two elements share one child and a set with two distinct equal-state children have different base pickles but identical HEAD hashes; `__tensor_unflatten__` can observe that topology.

File: test/dynamo/test_aot_autograd_cache.py [hunk 4]
```diff
@@ -4096,6 +4097,60 @@ def test_pickle_entry_strict_mode_raises(self):
...
+    def test_stabilize_nested_frozenset_in_set(self):
+        """frozenset elements inside a set must be recursively stabilized."""
+        gm = torch.fx.GraphModule({}, torch.fx.Graph())
+        pickler = AOTAutogradCachePickler(gm)
+
+        nested = {frozenset({"a", "b"}), frozenset({"c"})}
+        result = pickler._stabilize_tensor_subclass_metadata(nested)
+        self.assertIsInstance(result, _CanonicalSetMetadata)
+        self.assertEqual(result.container_type, set)
+        self.assertEqual(result.elements, tuple(sorted(result.elements)))
```
Comment (aorenste) [inline test/dynamo/test_aot_autograd_cache.py:4152]: [cleanup] My agent says:

> Assert the recursively canonicalized inner values:
>
> ```python
> expected = tuple(
>     sorted(
>         pickle.dumps(
>             _CanonicalSetMetadata(
>                 container_type=frozenset,
>                 elements=tuple(sorted(pickle.dumps(x) for x in inner)),
>             )
>         )
>         for inner in nested
>     )
> )
> self.assertEqual(result.elements, expected)
> ```
>
> The current assertions still pass if the recursive call is removed and each inner `frozenset` is pickled raw.

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

**Comment by Arindamsamanta2004** [comment]
@aorenste I've addressed your cleanup comments (memo-preserving sort plus the stronger nested test). Could you please approve the pending workflow runs when you get a chance? Thank you!

**Comment by Arindamsamanta2004** [comment]
The 2 failing jobs are unrelated trunk-unstable failures, not caused by this PR:

* `inductor_aoti_fallback (1,2)`: `test_aoti_debug_printer_codegen_cpu_with_stack_allocation` — missing `before_launch - aoti_torch_cpu_addmm_out` in generated code
* `inductor_aoti_fallback (2,2)`: `test_aoti_profiler_multi_output_fallback_input_shapes_cpu_with_stack_allocation` — codegen regex mismatch

This PR only touches the AOTAutograd cache-key pickler (`autograd_cache.py`) + dynamo tests. It cannot change AOTInductor C++ codegen strings. Same 2 tests failed on the previous head too, and both jobs are marked `unstable`.
